### PR TITLE
Lecture8

### PR DESCRIPTION
**Pull request by pliskov**
added lecture8 hometask implementation

### CODE REVIEW

File: Lecture8/src/test/java/AddRandomTractorToCartTest.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.concurrent.TimeUnit;
+
+public class AddRandomTractorToCartTest {
+    private final String homePageUrl = "https://onliner.by";
+    private final String shopCartUrl = "https://cart.onliner.by";
+    private final Browser browser = Browser.CHROME;
+
+    @Test
+    private void addRandomTractorToCartTest() throws InterruptedException {
+        WebDriver driver = WebdriverFactory.createWebdriver(browser);
```
Comment (acherkas) [inline Lecture8/src/test/java/AddRandomTractorToCartTest.java]: extract to @BeforeClass

File: Lecture8/src/main/java/Shop.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.PageFactory;
+import pageObject.*;
+
+public class Shop {
+    public static String addRandomTractorToCart(String homepageUrl, WebDriver driver) throws InterruptedException {
+        driver.get(homepageUrl);
+        Thread.sleep(5000);
+        HomePage homePage = PageFactory.initElements(driver, HomePage.class);
+        homePage.openCatalogPage();
```
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: should return CatalogPage

File: Lecture8/src/test/java/AddRandomTractorToCartTest.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import constant.Browser;
+import factory.WebdriverFactory;
+import org.openqa.selenium.WebDriver;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.TimeUnit;
+
+public class AddRandomTractorToCartTest {
+    private final String homePageUrl = "https://onliner.by";
```
Comment (acherkas) [inline Lecture8/src/test/java/AddRandomTractorToCartTest.java]: move to method open()

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    private List<WebElement> listAddToCartLinks;
+
+    public TractorPricesPage(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(this.driver, this);
+    }
+
+    public void addTractorToCart() throws InterruptedException {
+        sortOffersByPriceLink.click();
+        Thread.sleep(2000);
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/TractorPricesPage.java]: change to waiters

File: Lecture8/src/main/java/Shop.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.PageFactory;
+import pageObject.*;
+
+public class Shop {
+    public static String addRandomTractorToCart(String homepageUrl, WebDriver driver) throws InterruptedException {
```
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: remove static

File: Lecture8/src/main/java/Shop.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        driver.get(homepageUrl);
+        Thread.sleep(5000);
```
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: change to wait

File: Lecture8/src/main/java/Shop.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        HomePage homePage = PageFactory.initElements(driver, HomePage.class);
+        CatalogPage catalogPage = homePage.openCatalogPage();
+        CatalogTractorsPage catalogTractorsPage = catalogPage.openCatalogTractorsPage();
+        TractorPricesPage tractorPricesPage = catalogTractorsPage.openRandomTractorPricesPage();
+        tractorPricesPage.addTractorToCart();
+        return catalogTractorsPage.getSelectedTractor();
+    }
+
+    public static boolean isInCart(String shopCartUrl, String product, WebDriver driver) throws InterruptedException {
```
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: remove static

File: Lecture8/src/main/java/Shop.java (hunk 5 above)
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: remove homepageUrl

File: Lecture8/src/main/java/Shop.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.PageFactory;
+import pageObject.*;
+
+public class Shop {
+    public static String addRandomTractorToCart(String homepageUrl, WebDriver driver) throws InterruptedException {
+        driver.get(homepageUrl);
+        Thread.sleep(5000);
+        HomePage homePage = PageFactory.initElements(driver, HomePage.class);
```
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: HomePage hp= new HomePage(driver);
hp.open();
DiffPage df = hp.goTo....

File: Lecture8/src/main/java/pageObject/AbstractPage.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package pageObject;
+
+import org.openqa.selenium.WebDriver;
+
+public class AbstractPage {
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/AbstractPage.java]: rename to BasePage

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+import java.util.List;
+
+public class TractorPricesPage extends AbstractPage {
+
+    CatalogTractorsPage catalogTractorsPage = new CatalogTractorsPage(driver);
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/TractorPricesPage.java]: remove

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    private int selectedRandomNumber;
+    private String selectedTractor;
+
+    public CatalogTractorsPage(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(this.driver, this);
+    }
+
+    public int getSelectedRandomNumber() {
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: remove

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+public class CatalogTractorsPage extends AbstractPage {
+
+    @FindBy(xpath = "//div[@class='schema-product__group']//div[@class='schema-product__title']/a/span")
+    private List<WebElement> listTractorsTitles;
+
+    @FindBy(xpath = "//a[@class='schema-product__button button button_orange']")
+    private List<WebElement> listOffersButtons;
+
+    private int selectedRandomNumber;
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: remove

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    private String selectedTractor;
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: remove

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    public CatalogTractorsPage(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(this.driver, this);
+    }
+
+    public int getSelectedRandomNumber() {
+        return selectedRandomNumber;
+    }
+
+    public String getSelectedTractor() {
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: remove

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        return selectedTractor;
+    }
+
+    public TractorPricesPage openRandomTractorPricesPage() {
+        selectedRandomNumber = Helper.getRandom(listTractorsTitles.size());
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: remove

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        selectedTractor = listTractorsTitles.get(selectedRandomNumber).getText();
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: remove

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+    @FindBy(xpath = "//span[contains(.,'По цене')]/..")
+    private WebElement sortOffersByPriceLink;
+
+    @FindBy(xpath = "//td/a[@data-href='https://cart.onliner.by']")
+    private List<WebElement> listAddToCartLinks;
+
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/TractorPricesPage.java:19]: add title of tractor for webelement

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    public TractorPricesPage(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(this.driver, this);
+    }
+
+    public void addTractorToCart() throws InterruptedException {
+        sortOffersByPriceLink.click();
+        Thread.sleep(2000);
+        listAddToCartLinks.get(catalogTractorsPage.getSelectedRandomNumber()).click();
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/TractorPricesPage.java]: click random

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    }
```
Comment (acherkas) [inline Lecture8/src/main/java/pageObject/TractorPricesPage.java:42]: add method getTractorName() that returns parsed name of tractor

File: Lecture8/src/main/java/Shop.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public String addRandomTractorToCart() throws InterruptedException {
+        HomePage homePage = new HomePage(driver);
+        homePage.open();
+        CatalogPage catalogPage = homePage.openCatalogPage();
+        CatalogTractorsPage catalogTractorsPage = catalogPage.openCatalogTractorsPage();
+        TractorPricesPage tractorPricesPage = catalogTractorsPage.openRandomTractorPricesPage();
+        tractorPricesPage.addTractorToCart();
+        return catalogTractorsPage.getSelectedTractor();
```
Comment (acherkas) [inline Lecture8/src/main/java/Shop.java]: return tractorPricesPage.getTractorName();

File: Lecture8/src/test/java/AddRandomTractorToCartTest.java (hunk 1 above)
Comment (pliskov) [reply Lecture8/src/test/java/AddRandomTractorToCartTest.java]: extracted

File: Lecture8/src/main/java/Shop.java (hunk 2 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: done

File: Lecture8/src/test/java/AddRandomTractorToCartTest.java (hunk 3 above)
Comment (pliskov) [reply Lecture8/src/test/java/AddRandomTractorToCartTest.java]: moved

File: Lecture8/src/main/java/Shop.java (hunk 5 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: removed

File: Lecture8/src/main/java/Shop.java (hunk 6 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: implementation was changed to driver setting:
driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);

File: Lecture8/src/main/java/Shop.java (hunk 7 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: removed

File: Lecture8/src/main/java/Shop.java (hunk 5 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: removed

File: Lecture8/src/main/java/Shop.java (hunk 8 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: done

File: Lecture8/src/main/java/pageObject/AbstractPage.java (hunk 9 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/AbstractPage.java]: renamed

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java (hunk 10 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/TractorPricesPage.java]: removed

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java (hunk 11 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: removed

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java (hunk 12 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: removed

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java (hunk 13 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: removed

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java (hunk 14 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: removed

File: Lecture8/src/main/java/pageObject/CatalogTractorsPage.java (hunk 16 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/CatalogTractorsPage.java]: removed

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java (hunk 17 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/TractorPricesPage.java:19]: added

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java (hunk 18 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/TractorPricesPage.java]: click on cheapest

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java (hunk 19 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/TractorPricesPage.java:42]: added

File: Lecture8/src/main/java/Shop.java (hunk 20 above)
Comment (pliskov) [reply Lecture8/src/main/java/Shop.java]: fixed

File: Lecture8/src/main/java/pageObject/TractorPricesPage.java (hunk 4 above)
Comment (pliskov) [reply Lecture8/src/main/java/pageObject/TractorPricesPage.java]: changed

**Pull request merged**